VeganVirus/VeganVirus/FruitThrowAction.cpp

```diff
@@ -0,0 +1,161 @@
+#include "FruitThrowAction.h"
+
+FruitThrowAction::FruitThrowAction(Draw* draw, MouseManager& mouseManager, AudioManager& audioManager) : 
+	Action(FRUIT_THROW_REQ, FRUIT_THROW_ICON), _draw(draw), _mouseManager(mouseManager), _audioManager(audioManager),
+	_mouseDisabledAnimation(draw, FRAME_DURATION, MOUSE_ANIMATION_PATHS, MOUSE_DISABLED_ANIMATION_SIZE, MOUSE_DISABLED_ANIMATION_SIZE)
+{
+	// initialize trail
+	for (int i = 0; i < TRAIL_LENGTH; i++)
+	{
+		_trailPoints.push_back({ 0, 0 });
+	}
+}
+
+FruitThrowAction::~FruitThrowAction()
+{
+	for (FruitThrow* fruit : this->_currentFruits)
+	{
+		delete fruit;
+	}
+}
+
+void FruitThrowAction::start()
+{
+	this->_fruitLeft += FRUIT_COUNT;
+	this->_timer = 0;
+
+	// set previous mouse position
+	Point_t point = _mouseManager.getMousePosition();
+	_px = point.x;
+	_py = point.y;
+}
+
+void FruitThrowAction::update(double dt)
+{
+	if (this->_fruitLeft == 0 && this->_currentFruits.size() == 0)
+		return;
+
+	double x1 = _px, y1 = _py;
+
+	if (!_mouseDisabled)
+	{
+		// get new mouse position
+		Point_t point = _mouseManager.getMousePosition();
+		// calc velocity
+		_vx = (point.x - _px) / dt;
+		_vy = (point.y - _py) / dt;
+	}
+
+	// compute subframes
+	double t = dt;
+	for (; t > COMPUTATION_SUB_FRAME_INTERVAL; t -= COMPUTATION_SUB_FRAME_INTERVAL)
+	{
+		x1 += _vx * COMPUTATION_SUB_FRAME_INTERVAL;
+		y1 += _vy * COMPUTATION_SUB_FRAME_INTERVAL;
+		this->subFrameUpdate(COMPUTATION_SUB_FRAME_INTERVAL, x1, y1);
+	}
+	x1 += _vx * t;
+	y1 += _vy * t;
+	this->subFrameUpdate(t, x1, y1);
+
+	_px = x1;
+	_py = y1;
+	_trailPoints.push_back({ (int)x1, (int)y1 });
+	_trailPoints.pop_front();
+
+	if (_mouseDisabled)
+	{
+		_mouseManager.setMousePosition(x1, y1);
+
+		// update mouse freeze
+		if (_mouseDisabledTimer > 0)
+		{
+			_mouseDisabledTimer -= dt;
+			this->_mouseDisabledAnimation.update(dt, x1 - MOUSE_DISABLED_ANIMATION_SIZE / 2, y1 - MOUSE_DISABLED_ANIMATION_SIZE / 2);
+			if (_mouseDisabledTimer <= 0)
+			{
+				_mouseDisabled = false;
+				BlockInput(FALSE);
+			}
+		}
+	}
+
+	for (FruitThrow* fruit : this->_currentFruits)
+	{
+		fruit->draw();
+	}
+
+	this->_timer -= dt;
+	if (this->_fruitLeft && this->_timer <= 0)
+	{
+		this->_currentFruits.push_back(new FruitThrow(this->_draw, this->_mouseManager));
+		this->_timer += (float)rand() / RAND_MAX * (MAX_FRUIT_TIMER - MIN_FRUIT_TIMER) + MIN_FRUIT_TIMER;
+		this->_fruitLeft--;
+	}
+
+	// draw trail
+	TrailPoint prev = _trailPoints.front();
+	int width = 2;
+	for (auto it = ++_trailPoints.begin(); it != _trailPoints.end(); it++)
```

Comment:
please don't use `auto`. it can cause problem

Reply:
`auto` keyword is widely used for iterators. I disagree that it can cause problems, at least in this situation.